cumulusci/robotframework/pageobjects/PageObject.py

```diff
@@ -0,0 +1,125 @@
+import re
+import os.path
+import sys
+import logging
+from robot.libraries.BuiltIn import BuiltIn
+
+
+class PageObject(object):
+    ROBOT_LIBRARY_SCOPE = "GLOBAL"
+
+    # These must be defined by subclasses
+    PAGE_TYPE = None
+    OBJECT_API_NAME = None
+
+    """Base class to use for all page objects
+
+    Derived classes should define PAGE_TYPE based on the appropriate
+    pageReference type but with the "standard_" prefix removed (eg:
+    objectPage rather than standard_objectPage)
+
+    This class provides the following properties for accessing
+    keywords in other libraries:
+
+    builtin    - robot's built-in keywords (eg: self.builtin.log("..."))
+    selenium   - SeleniumLibrary keywords
+    cumulusci  - cumulusci.robotframework.Salesforce library
+    salesforce - cumulusci.robotframework.CumulusCI library
+
+    Derived classes can also define the following special methods:
+
+    _go_to_tab
+
+      called by the Salesforce keyword 'go to tab'. The default
+      implementation will used the PAGE_TYPE and OBJECT_API_NAME
+      attributes to construct a URL to the page. You can override
+      this method to go to any URL you want.
+
+    _is_current_page
+
+      called by the keyword 'current page should be'. By default
+      it will get the current page location and compare it to a
+      URL based on the PAGE_TYPE and OBJECT_API_NAME attributes.
+
+    When this library is imported it will be added to PYTHONPATH, which is
+    what python uses to resolve name conflicts. If your page object defines
+    a keyword with the same name as some other page object, you can control
+    which one has precedence based on the order that you import the libraries.
```

Comment:
Wait, I thought we were controlling the precedence using `Set Library Search Order`; what's the Python path adjustment needed for? Monkeying with the Python path sounds scary, and also this seems error-prone if some other test suite already imported some of the page objects in a different order.

Reply:
Your observations are correct.

Part of the reason it took me so long to push code was that I was trying to get to a place where a test should be able to import a page with `Library  SurfingSafariHomePage` in their settings, without having to tell robot explicitly where to look.  This code is a remnant of that, and is no longer necessary. I should have removed it before pushing the code.

For the moment at least, you have to fully qualify the page objects when you import them (see for example the file `SurfingSafari.resource`). 

One solution to avoid having to fully qualify the library when importing is to add some something to cumulusci.yml that tells robot where to look for libraries. For this first cut I didn't want to do that.

I'm curious what you think about the goal of being able to import using just the base name (eg: `SurfingSafariHomePage`) versus a fully qualified path (eg: `robot/SurfingSafari/resources/SurfingSafariHomePage.py`).

Reply:
My gut feeling is that importing the full path is probably better. That was someone reading a test suite can just go to that file instead of needing to understand what the resolution rules are.